### PR TITLE
luci-app-watchcat: add support for new features + JS refactor

### PR DESCRIPTION
**Pull request by nickberry17**
Signed-off-by: Nicholas Smith <nicholas@nbembedded.com>

- Add support for restarting an interface (including ModemManager interface support)
- support for pinging via an interface
- support for specifying ping packet size
- Refactor/rewriting some of the user facing text (will require a re-translation)
- Rewrite in JavaScript

Depends on: https://github.com/openwrt/packages/pull/14225

### CODE REVIEW

**Comment by nickberry17** [comment]
https://github.com/openwrt/packages/pull/14225 has been merged

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:121]: Please convert the file to tabs

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:112]: I think the depends could be one liner an does not need to be split up in 3 lines. Please have a look to the others as well in this line

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:90]: What does the commentar `sm` mean?

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:13]: translate missing

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:31]: The second translate into a new line and the splitting of the strings with `"<message>" ..` fits.
But the whole thing should be consistent in the whole file.

**Review by feckert — CHANGES_REQUESTED** [review]

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (nickberry17) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua]: Like this?
```
pinghost:depends({ mode = "ping_reboot" })
pinghost:depends({ mode = "restart_iface" })
```

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (nickberry17) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua]: Do you mean the extra long line at the end?  I've reformatted it so they are all approximately the same length.

Comment (nickberry17) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua]: Also, should the translation strings include HTML tags, or should we take them out?  We could do something like this if it makes more sense:

`"<b>" .. translate("Hello") .. "</b>"`

**Comment by kofec** [comment]
New watchcat was mergerd so maybe we can also merge this changes ?

**Comment by hnyman** [comment]
> maybe we can also merge this changes ?

At least not when there are conflicts. (apparently some translations have changed while this has been open)

> > Conflicting files
> > applications/luci-app-watchcat/po/zh_Hant/watchcat.po

**Comment by nickberry17** [comment]
> > maybe we can also merge this changes ?
> 
> At least not when there are conflicts. (apparently some translations have changed while this has been open)
> 
> > > Conflicting files
> > > applications/luci-app-watchcat/po/zh_Hant/watchcat.po

Hey, I have fixed the conflict

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [inline applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:19]: The TropDown gets quite long due to the translation. Can we not shorten this.

**Review by feckert — CHANGES_REQUESTED** [review]
@nickberry17 Sorry for the delay
Can you please indent the translate() function consistently like in the other modules?
So that we have the following.

```
m = Map("system",
       translate("Watchcat"),
       translate("Here you can set up several checks and actions to take in the event that a host becomes unreachable. " ..
                 "Click the <b>Add</b> button at the bottom to set up more than one action."))
```
You can use the old code as an example.
https://github.com/openwrt/luci/pull/4658/files#diff-e5e1125156a1388d7a259645b806b12c12e6acdf221530c7ef581156251fcacfL30

Pay attention to the tabs at the beginning of the line. You can then fill the rest of the line with space until the indentation fits.

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (feckert) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua]: There is no default. Personally, I think it's better if you're not there. But everyone does that differently here in the LuCI.

File: applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua
Comment (nickberry17) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:19]: Thanks for the feedback, will do. It just looks this way at the moment because of the auto formatter I have in my editor, which I need to fixup.

Comment (nickberry17) [reply applications/luci-app-watchcat/luasrc/model/cbi/watchcat/watchcat.lua:19]: > The TropDown gets quite long due to the translation. Can we not shorten this.

Yes, I'll rephrase and try to keep the option descriptions from getting too long by putting all of the context in the section's description.

**Comment by nickberry17** [comment]
Hey, I have rewritten it using the JavaScript API to get rid of the luci-compat dependency.

**Comment by feckert** [comment]
From my side, this looks good. However, I still have one question. What happened to the other translations? Why are they deleted? Shouldn't they also be updated with the new content?

**Comment by nickberry17** [comment]
> From my side, this looks good. However, I still have one question. What happened to the other translations? Why are they deleted? Shouldn't they also be updated with the new content?

My mistake.  I have fixed that now

File: applications/luci-app-watchcat/root/etc/uci-defaults/40_luci-watchcat
Comment (feckert) [inline applications/luci-app-watchcat/root/etc/uci-defaults/40_luci-watchcat:1]: Sorry if I ask again. Are you sure we don't need this file? As far as I can see, the service has not yet been switched to procd. To trigger an restart or reload on config change.
Or is this created automatically now.
I am unfortunately no longer quite sure how it is now

Comment (nickberry17) [reply applications/luci-app-watchcat/root/etc/uci-defaults/40_luci-watchcat]: Hey, you were right to point that out.  I needed to make some changes to get it working with ucitrack.  https://github.com/openwrt/packages/pull/14569

Comment (feckert) [reply applications/luci-app-watchcat/root/etc/uci-defaults/40_luci-watchcat]: I would prefer if we could switch the service to procd. Then we would no longer have the annoying ucitrack.

Comment (nickberry17) [reply applications/luci-app-watchcat/root/etc/uci-defaults/40_luci-watchcat]: Migration to procd has been done

**Comment by feckert** [comment]
I think we are still waiting for the [P/R](
https://github.com/openwrt/packages/pull/14569), then the reload will also work when the configuration is changed.

**Review by feckert — APPROVED** [review]
Looks good so far. But do not do a run test until the `watchcat` packages get merged.

**Comment by nickberry17** [comment]
ping @feckert

File: applications/luci-app-watchcat/htdocs/luci-static/resources/view/watchcat.js
Comment (feckert) [inline applications/luci-app-watchcat/htdocs/luci-static/resources/view/watchcat.js:13]: @nickberry17 thanks for the ping again. I have a lot to do at the moment and can't keep up with my work.

Why do we need **stats** here? I don't see you using that.
I therefore think the **load** function is superfluous and therefore also the **'require fs'** include.

You also have to do a rebase so that the translation fits again, otherwise I can't merge it.

Comment (nickberry17) [reply applications/luci-app-watchcat/htdocs/luci-static/resources/view/watchcat.js]: @feckert No pressure, I know how it can be sometimes.  I rebased, updated the ACL and refreshed the translation files and just ran a quick test on the latest master.  Are we ready to roll?

**Comment by feckert** [comment]
Sorry if I'm being annoying, I think this is ready to merge.
But could you still stash the commits so we have one?

**Comment by nickberry17** [comment]
> Sorry if I'm being annoying, I think this is ready to merge.
No, you're fine, I appreciate the review.  
> But could you still stash the commits so we have one?
Stash or squash?

**Comment by nickberry17** [comment]
Squashed

**Comment by feckert** [comment]
thanks merged

**Comment by kofec** [comment]
Hi, are you going to backport this changes to 21.02 ?

**Comment by nickberry17** [comment]
> Hi, are you going to backport this changes to 21.02 ?

I will if that is allowed
ping @feckert

**Comment by feckert** [comment]
From my point of view, that's fine.
The openwrt-21.02 branch has not yet been released.
It is still in the stabilization phase.
I would suggest that one of you @nickberry17 @kofec  creates a pull request with the cherry-picked commits.
Please do not forget the package feed ;-) commits

**Comment by ChandraOrbit** [comment]
suggestion
on the Mode: Restart Interface
to check if the wireless is in AP or STA mode,
So if we change the wireless interface from STA to AP and we forget to set the watchcat, the watchcat will automatically deactivate the interface restart mode.

**Comment by kofec** [comment]
@nickberry17 Can you please add these changes to 21.02 as you probably have environment ready  (from trunk)

**Pull request merged**
